### ISSUE TITLE
[1.0b + eventlet] cookies and sessions fall out of sync with main site thread

### ISSUE THREAD

**Issue by ktoh** [miguelgrinberg/Flask-SocketIO#148]
I'm running across the following behavior with 1.0b + eventlet:

1. Start app
2. flask-socketio creates a request context for websockets
3. App sets a cookie and/or session variable (e.g. a user logs in)
4. The new cookie/session variables are visible to regular web request objects via `flask.request.cookies` and `flask.session` 
4. However, these same variables are not visible to the socket server, i.e. `flask.request.cookies` and `flask.session` point to stale objects inside the flask-socketio request context.

I've created a simple app to reproduce the issue here:  
* [app.py](https://gist.github.com/ktoh/333fb96019832b251565)  
* [index.html](https://gist.github.com/ktoh/f062a67c546e88035652)

This is causing a few problems because various flask libraries and many apps use cookies and session variables to keep track of user login status, session parameters, etc.  

For example, `current_user` (in flask-security/flask-login) depends on cookies, so if a user shows up and logs in via web, the websocket request context will not recognize the user as logged in because `current_user` is loading from stale cookie/session data.

**Comment by miguelgrinberg** [comment]
This behavior is consistent with the older versions. There is a note in the documentation regarding to this:

> The session context global behaves in a different way than in regular requests. A copy of the user session at the time the SocketIO connection is established is made available to handlers invoked in the context of that connection. Any changes made to the session inside a SocketIO handler are preserved, but only in the SocketIO context, these changes will not be seen by regular HTTP handlers.

I think I should expand on this to indicate that changes made in HTTP requests after the socket connection was made are not accessible in the socket handlers.

The documentation also has a note regarding using server-side sessions to avoid these problems. That note is not accurate, I either need to remove it, or make sure server side sessions do allow sharing of values between HTTP and socket handlers.

**Comment by ktoh** [comment]
Thanks @miguelgrinberg. There does seem to be something different with 0.6 because the previous code was able to load current_user correctly (via flask login, via cookie) if a user logged in during the session. In other words, `request.cookies` worked inside the socket server context. 

Any ideas how this might have happened?

**Comment by miguelgrinberg** [comment]
Okay, I'll try your example with the old and the new versions and figure out what's going on.

**Comment by ktoh** [comment]
If it helps, I'd guess that the issue is with `request.cookies`.  I'm using server-side sessions (redis), so all that needs to happen is for the socket handler and the http servers to have access to a single session ID `sid` cookie variable for them to stay in sync.

This seemed to work fine out of the box with 0.60 and prior versions, but `request.cookies` now seems to return different values between the two threads.

I see that `request_context` is applied differently between [version 0.60](https://github.com/miguelgrinberg/Flask-SocketIO/blob/master/flask_socketio/__init__.py#L153) and [version 1.0x](https://github.com/miguelgrinberg/Flask-SocketIO/blob/v1.0/flask_socketio/__init__.py#L111) so I'll start my own investigation there.

**Comment by miguelgrinberg** [comment]
Ah, okay, that makes more sense. Server side sessions do appear to be broken in 1.0, I realized that while looking at the docs and the code, which do not agree.

**Comment by ktoh** [comment]
OK on further investigation it appears that the root cause here is client side, not server side.

For posterity, here is the issue:

* Flask-SocketIO creates a convenient request context for the websocket handlers, so that developers can use flask constructs like `session`, `current_user`, `requests`, etc.
* It does this by taking the request which created the HTTP websocket (i.e. the connect/upgrade request), and using that to create the context.
* For most clients (i.e. most browsers), the connect/upgrade request includes a cookie header, which allows cookies to be passed onto the request context for socket handlers.   This in turn allows other flask constructs like `current_user` and `session` to work because they may depend on cookie data for identifying the current session, current user, etc.
* But, **you cannot assume browsers will pass cookies during the connect/upgrade request**, or during any other websocket-related request.  
   * A discussion and a link to the RFC specification is [here](https://groups.google.com/forum/#!msg/faye-users/naVeKOkiYb4/QtUMfhk0bqoJ)

If your client (in my case, an iOS app) decides they don't want to send cookies along with the request, then various flask constructs may fail because they will be unable to load state from the non-existent cookie data.

I'm still working through a solution and will update this response, but in the meantime:

* A discussion of the issue on socket.io github is [here](https://github.com/socketio/socket.io-client/issues/648#issuecomment-42154344)
* One approach is to pass relevant cookie/session data explicitly to the from the client to the server after the websocket connection is established, as discussed [here](http://stackoverflow.com/questions/4753957/socket-io-authentication).  

@miguelgrinberg, unless you can think of a way to synchronize cookies as request objects get updated in the http thread (this may create all kinds of problems with distributed servers), I'd suggest removing the bug tag on this issue and just updating the docs on how  request_context is established, as you note above.




It turns out that cookies [MAY OR MAY NOT](https://groups.google.com/forum/#!msg/faye-users/naVeKOkiYb4/QtUMfhk0bqoJ) be passed ont

**Comment by miguelgrinberg** [comment]
Oh, that's interesting. What's unclear to me, is how the server can match a socket client with an HTTP client without cookies. There's really nothing else that cannot be faked that can be used as identification.

Which client are you using, the Swift client? Did you check if the iOS client can be given optional headers to pass on the request?

**Comment by ktoh** [comment]
@miguelgrinberg I'm using react-native on iOS, but this seems to be an iOS issue rather than a react issue.   BTW, flask-socketio otherwise seems to work well with React-Native.

It looks like the socket.io community doesn't want to allow headers to be set explicitly, so the approach I'm exploring is passing the cookie content (really the session ID is all that's necessary) via the connect/upgrade URL using a GET argument.

That's not ideal, but since the request is going over HTTPS that should be no different security-wise from using a cookie header.  Wiring up the server-side sessions back to that non-cookie ID, on the other hand, may be a bit tricky.

**Comment by miguelgrinberg** [comment]
@ktoh Maybe you can get away with callilng `app.open_session` from your handler function. This method takes the request as an argument, you may be able to inject the cookie in the request object and fool the session code this way.

**Comment by ktoh** [comment]
@miguelgrinberg it looks like flask opens the session first thing upon receiving the request, and i'm a bit worried about injecting state haphazardly.

Two possible solutions here:
1. It looks like there's a forthcoming update of engine.io which fixes this inconsistency around headers [here](https://github.com/socketio/engine.io-client/pull/379).  I think this is the right approach since it really breaks isomorphic architecture to have different clients connect differently over the same (http) protocol, and headers are used all the time for transporting context so why should socket.io behave differently.

2. Write request middleware to inject the cookie data back into the request before it hits the session opener.

The idea of #2 makes me throw up in my mouth so I'm considering pulling down the master branch of socket.io and using the pre-release code.  The code has been merged and it's just going through testing so I think it's pretty safe to assume it's going to be supported.

Will report back if I get it to work, because I expect this is going to be an issue for others writing for multi-client (web/mobile/etc) and expecting `current_user` and `session` to work reasonably well.

**Comment by ktoh** [comment]
Here's my post-mortem on how I resolved the issue.

### Approaches considered:

1. **Wait for socketio/engineio release to fix this**.  There is a forthcoming release ([here](https://github.com/socketio/engine.io-client/pull/379)) of engine.io which allows for headers to be added to socket requests.  But:  (a) it hasn't been released yet; (b) even when it does, who knows how long it'll take socket.io to pick up the update and then enable custom headers.

2. **Write custom middleware to inject the cookie data back into the request**.  This is certainly viable.  The middleware could take cookie content from a query string or other non-header data (e.g. POST content) and inject it into a cookie header before passing it onto flask.

3. **Pass the cookie/session data explicitly from client to server after the initial websocket connection is established**.  This approach is outlined [here](http://stackoverflow.com/a/4754806/1516199).  I didn't like the back-and-forth dialog between client-server because it just seemed unnecessary.

4. **Create a smarter session object which can recognize websocket connections and load session accordingly**.  This was the approach that worked best for me, because I already had a custom server-side session manager (based on [this](http://flask.pocoo.org/snippets/75/)) so this would minimize the number of hooks I have into Flask.

### How it's done

* **Client side**.  Pass the session ID and any other relevant data as query strings during the initial websocket `connect`:

```
var io = require('socket.io-client/socket.io')
_socket = io.connect(
    'mywebsite.com/socket.io?session=' + mysessionvariable,
    { 'jsonp': false,  'reconnect': true }
  )
```

* **Server side**.  When the session object starts a session, it calls [`open_session(request)`](http://flask.pocoo.org/docs/0.10/api/#flask.Flask.open_session).  Simply check the request to see if it's a websocket upgrade, and if so then set the session ID from the query string:

```
if request.headers.get('Connection', "").lower() == 'upgrade':
    # Retrieve session ID from URL query string
    sessionid = request.args.get('session', None)
else:
    # Retrieve session ID from cookie
    sessionid = request.cookies.get('session', None)
```

This allows for the session to be re-established by the socket handler just once, on the initial connect, without any need for additional back and forth dialog.

**Comment by miguelgrinberg** [comment]
Fantastic write up, thanks a lot. I will see how I can incorporate these ideas into the documentation, as I'm sure this is going to come up again.

**Comment by ktoh** [comment]
I agree this should go into docs for now rather than attempting some kind of code solution.
One of the great things about flask_socketio is its pythonic simplicity. 
By trying to solve this problem inside the library, we risk having it become too opinionated and frankly off-spec since we would essentially be getting ahead of where both socket.io and the RFC spec are today.

**Comment by miguelgrinberg** [comment]
I believe I have addressed the difference in behavior with sessions between 0.x and 1.x releases. This fix is in version 1.0b4. Can you please confirm it works for you?

The fix is really simple, this copied session that exists for Socket.IO handlers is only created if these handlers modify the session. As long as the socket handlers treat the session as read-only, the original session contents will be used. If you use server-side sessions, that means that any updates to the session done by regular HTTP route handlers will be accessible to socket handlers. But, the moment a socket handler modifies the session, a copy of the session will be created and use in all future socket handler calls. I hope this makes sense.
